rfc/0000-batch-cancellation.rst

```diff
@@ -0,0 +1,538 @@
+ready cores and jobs upon cancelling the batch by subtracting the
+number of **cancellable** resources and jobs for that batch. This way
+both the autoscaler and scheduler instantly see this change in demand
+with their respective fair share algorithms. In addition, the
+autoscaler will not provision unnecessary resources for the cancelled
+batch as the query that scans the head of the jobs table will ignore
+jobs that are "Ready" but in batches that have been
+cancelled. Likewise, the SQL query the scheduler uses will ignore
+"Ready" jobs for batches that have been cancelled and the fair share
+algorithm will instantly adapt to the change in demand from the
+cancelled batch to schedule other users and batches fairly.
+
+More concretely, there are two tables that keep track of resources:
+
+- ``user_inst_coll_resources``: Keeps track of the total number of
+  jobs and cores for each user and instance collection for jobs in
+  Ready, Running, and Creating states as well as the number of
+  cancelled jobs in each state. Note this table is not parameterized
+  by batch ID.
+- ``batch_inst_coll_cancellable_resources``: Keeps track of the total
+  number of **cancellable** jobs and cores for jobs in Ready, Running,
+  and Creating states. Note this table is parameterized by
+  batch id. The numbers in this table do not include counts for
+  **always_run** jobs as they cannot be cancelled.
+
+These two tables are tokenized for fast concurrent
+updates. Tokenization means that the value for a single key in the
+database is represented by up to 200 rows in order to avoid
+serialization of updates. Therefore, to compute the number of jobs or
+free cores, the query must do an aggregation for all rows for a given
+key which can be up to 200 rows.
+
+These two database tables are kept up to date with a trigger on the
+``jobs`` table after updates. This trigger increments and decrements
+the corresponding rows in these tables based on the previous job state
+and the new job state. The code in the trigger is extremely
+complicated as it has to take into account the current job state,
+whether the batch has been cancelled, whether the job has been
+cancelled (failed parent dependencies), and whether the job is
+**always_run**. This trigger is currently ~170 lines and contains many
+nested if-else statements that are easy to make mistakes on when
+refactoring the code.
+
+When a batch is cancelled by either the user or the system
+("cancel_after_n_failures" or billing limits have been reached), the
+following occurs:
+
+1. A row for that batch id is inserted into a table that keeps track
+   of the batches that have been cancelled.
+2. The number of **cancellable** ready cores and number of ready jobs
+   for that batch are calculated by aggregating over the
+   ``batch_inst_coll_cancellable_resources`` table and then
+   subtracting those counts from the corresponding values in the
+   ``user_inst_coll_resources`` table.
```

Comment:
This has been a very enlightening read, thank you for putting the time and effort into this. I have been confused until now why fair share had anything to do with cancellation and this bullet clarifies that point entirely. Nevertheless, I am unconvinced that these two topics need to be related and do not yet understand the true motivation for performing this subtraction.

In my mind, a fair share algorithm need only concern itself with the resources currently in use, i.e. currently Running jobs. As the determiner of fair share, why would I care if a job that's running belongs to a cancelled batch? Is that user not still hogging those resources until the jobs themselves are cancelled?

Is the following scenario a correct understanding of the current system?

Two users, A and B, are perfectly sharing the cluster 50/50, each running one batch. User A submits a second, identical, batch and cancels their first batch. Immediately following cancellation, individual jobs have not yet been cancelled (and will be cancelled slowly) but the scheduler sees A as consuming 0 resources compared to B's resources that have not changed, so the scheduler will favor user A in scheduling jobs. This may result in user A using >50% of the cluster depending on the speed of the scheduler vs the canceler. As user B, I would feel pretty cheated in this scenario.

Reply:
I think I misunderstood earlier, does this not wipe out `Running` resources, rather wipes out some `Ready` resources? If that is the case, why does the scheduler care about how many Ready jobs a user has? I would assume that a fair share should be determined based on the currently `Running` jobs, and the scheduler just tries to select up to that number of ready jobs